osu.Game/Lists/IntervalList.cs

```diff
@@ -0,0 +1,128 @@
+        /// </summary>
+        /// <param name="value">The value to check for.</param>
+        public bool IsInAnyInterval(T value)
+        {
+            if (intervals.Count == 0)
+                return false;
+
+            // Clamp the nearest index in case there were intervals
+            // removed from the list causing the index to go out of range.
+            NearestIntervalIndex = Math.Clamp(NearestIntervalIndex, 0, Count - 1);
+
+            if (type_comparer.Compare(value, this[NearestIntervalIndex].End) > 0)
+            {
+                while (type_comparer.Compare(value, this[NearestIntervalIndex].End) > 0 && NearestIntervalIndex < Count - 1)
+                    NearestIntervalIndex++;
+            }
+            else
+            {
+                while (type_comparer.Compare(value, this[NearestIntervalIndex].Start) < 0 && NearestIntervalIndex > 0)
+                    NearestIntervalIndex--;
+            }
+
+            var nearestInterval = this[NearestIntervalIndex];
+
+            return type_comparer.Compare(value, nearestInterval.Start) >= 0 &&
+                   type_comparer.Compare(value, nearestInterval.End) <= 0;
+        }
+
+        /// <summary>
+        /// Adds a new interval to the list.
+        /// </summary>
+        /// <param name="start">The start value of the interval.</param>
+        /// <param name="end">The end value of the interval.</param>
+        public void Add(T start, T end) => Add(new Interval<T>(start, end));
+
+        #region ICollection<Interval<T>>
+
+        public int Count => intervals.Count;
+
+        bool ICollection<Interval<T>>.IsReadOnly => false;
+
+        /// <summary>
+        /// Adds a new interval to the list
+        /// </summary>
+        /// <param name="interval">The interval to add.</param>
+        public void Add(Interval<T> interval) => intervals.Add(interval);
+
+        /// <summary>
+        /// Removes an existing interval from the list.
+        /// </summary>
+        /// <param name="interval">The interval to remove.</param>
+        /// <returns>Whether the provided interval exists in the list and has been removed.</returns>
+        public bool Remove(Interval<T> interval) => intervals.Remove(interval);
+
+        /// <summary>
+        /// Removes all intervals from the list.
+        /// </summary>
+        public void Clear() => intervals.Clear();
+
+        public void CopyTo(Interval<T>[] array, int arrayIndex) => intervals.CopyTo(array, arrayIndex);
+
+        public bool Contains(Interval<T> item) => intervals.Contains(item);
+
+        public IEnumerator<Interval<T>> GetEnumerator() => intervals.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        #endregion
+
+        #region IReadOnlyList<Interval<T>>
+
+        public Interval<T> this[int index]
+        {
+            get => intervals[index];
+            set => intervals[index] = value;
+        }
+
+        #endregion
```

Comment:
why? who's going to ever use these?

Reply:
implementation of `ICollection<Interval<T>>` and `IReadOnlyList<Interval<T>>`, I guess what your review suggests is to make it so simple that the implementation becomes enough for its usages.